data-processing/common/upload_entities.py

```diff
@@ -309,3 +313,53 @@ def resolve_model(entity_identifier: EntityIdentifier) -> Optional[Entity]:
                     )
 
     return models
+
+
+def write_to_file(entity_infos: List[EntityUploadInfo], output_file_name: str) -> None:
+    # an attempt to make life easier if we change how we want to format this
+    FORMAT_VERSION = "v0"
+
+    logging.info(
+        "About to write %d entity infos to %s (version: %s).",
+        len(entity_infos),
+        output_file_name,
+        FORMAT_VERSION,
+    )
+    to_write = {
+        "version": FORMAT_VERSION,
+        "data": [dataclasses.asdict(entity_info) for entity_info in entity_infos],
+    }
+    already_exists_msg = (
+        f"File {output_file_name} already exists. "
+        + "Not overwriting. Entity infos will not be written."
+    )
+    assert not os.path.exists(output_file_name), already_exists_msg
+
+    with open(output_file_name, "w") as output_file:
+        json.dump(to_write, output_file)
+
+
+def save_entities(
+    s2_id: S2Id,
+    arxiv_id: ArxivId,
+    entity_infos: List[EntityUploadInfo],
+    data_version: Optional[int],
+    output_details: OutputDetails,
+    filename: str,
+) -> None:
+
+    if output_details.save_to_file():
+        logging.info("Saving to file...")
+        # should always be true, but let's just make sure
+        assert output_details.output_dir is not None, "Expected a defined output dir!"
```

Comment:
For these assertions, I am wondering whether they should be created as assertion messages, or as their own exception type. I recall russellr implemented error catchers in the data-pipeline runner code for common types of errors so that they could be surfaced as counters in DataDog. Do you know whether we would be able to count up assertion failures based on their messages from the pipeline runner code?

Reply:
Yep, I believe the way the pipeline runner code counts things up is by looking at what is printed out (except for a couple of cases where the count is based on exit code), so I think it should be able to count these up as well.

Sidenote: this assertion in particular is something that we shouldn't ever hit (because the way the input argument validation is set up, this should never happen), but sometimes I sprinkle things like this in to guard against code changes potentially changing the expectation that this would never happen.